File: 30_nestループ/python/main.py
#!/usr/bin/env python
# coding: -utf8


def table_printer(begin, do_work, end):
    begin()
    for col in range(13):
        do_work(col)
    end()


def print_header():
    table_printer(
        lambda: print("|    ", end=""),
        lambda col: print("|%4d" % col, end=""),
        lambda: print("|\n", end=""),
    )


def print_horizontal_line():
    table_printer(
        lambda: print("+----", end=""),
        lambda _: print("+----", end=""),
        lambda: print("+\n", end=""),
    )


def print_row(row):
    table_printer(
        lambda: print("|%4d" % row, end=""),
        lambda col: print("|%4d" % (row * col), end=""),
        lambda: print("|\n", end=""),
    )


if __name__ == '__main__':
    print_horizontal_line()
    print_header()
    for i in range(13):
        print_horizontal_line()
        print_row(i)
    print_horizontal_line()

File: 30_nestループ/go/main.go
package main

import "fmt"

func tablePrinter(begin func(), doWork func(int), end func()) {
	begin()
	for col := 0; col <= 12; col++ {
		doWork(col)
	}
	end()
}
func printHeader() {
	tablePrinter(
		func() { fmt.Printf("|    ") },
		func(col int) { fmt.Printf("|%4d", col) },
		func() { fmt.Printf("|\n") },
	)
}

func printHorizontalLine() {
	tablePrinter(
		func() { fmt.Printf("+----") },
		func(int) { fmt.Printf("+----") },
		func() { fmt.Printf("+\n") },
	)
}

func printRow(row int) {
	tablePrinter(
		func() { fmt.Printf("|%4d", row) },
		func(col int) { fmt.Printf("|%4d", row*col) },
		func() { fmt.Printf("|\n") },
	)
}

func main() {
	printHorizontalLine()
	printHeader()
	for i := 0; i <= 12; i++ {
		printHorizontalLine()
		printRow(i)
	}
	printHorizontalLine()
}
